stdpopsim/species.py

```diff
@@ -170,42 +182,112 @@ def get_contig(self, chromosome, genetic_map=None, length_multiplier=1):
-        if (chromosome is not None and
-                chromosome.lower() in ("x", "y", "m", "mt", "chrx", "chry", "chrm")):
-            warnings.warn(stdpopsim.NonAutosomalWarning(
+        if chromosome is not None and chromosome.lower() in (
+            "x",
+            "y",
+            "m",
+            "mt",
+            "chrx",
+            "chry",
+            "chrm",
+        ):
+            warnings.warn(
+                stdpopsim.NonAutosomalWarning(
                     "Non-autosomal simulations are not yet supported. See "
                     "https://github.com/popsim-consortium/stdpopsim/issues/383 and "
-                    "https://github.com/popsim-consortium/stdpopsim/issues/406"))
-        chrom = self.genome.get_chromosome(chromosome)
-        if genetic_map is None:
-            logger.debug(f"Making flat chromosome {length_multiplier} * {chrom.id}")
-            gm = None
-            recomb_map = msprime.RecombinationMap.uniform_map(
-                chrom.length * length_multiplier, chrom.recombination_rate)
-        else:
+                    "https://github.com/popsim-consortium/stdpopsim/issues/406"
+                )
+            )
+        if chromosome is None:
+            if genetic_map is not None:
+                raise ValueError("Cannot use genetic map with generic contic")
             if length_multiplier != 1:
-                raise ValueError("Cannot use length multiplier with empirical maps")
-            logger.debug(f"Getting map for {chrom.id} from {genetic_map}")
-            gm = self.get_genetic_map(genetic_map)
-            recomb_map = gm.get_chromosome_map(chrom.id)
-
-        ret = stdpopsim.Contig(
-            recombination_map=recomb_map, mutation_rate=chrom.mutation_rate,
-            genetic_map=gm)
+                raise ValueError("Cannot use length multiplier for generic contig")
+            if inclusion_mask is not None or exclusion_mask is not None:
+                raise ValueError("Cannot use mask with generic contig")
+            if length is None:
+                raise ValueError("Must specify sequence_length of generic contig")
+            non_autosomal_lower = ["x", "y", "m", "mt", "chrx", "chry", "chrm"]
+            L_tot = 0
+            r_tot = 0
+            u_tot = 0
+            for chrom_data in self.genome.chromosomes:
+                if chrom_data.id.lower() not in non_autosomal_lower:
+                    L_tot += chrom_data.length
+                    r_tot += chrom_data.length * chrom_data.recombination_rate
+                    u_tot += chrom_data.length * chrom_data.mutation_rate
+            u = u_tot / L_tot
+            r = r_tot / L_tot
+            recomb_map = msprime.RecombinationMap.uniform_map(length, r)
+            ret = stdpopsim.Contig(recombination_map=recomb_map, mutation_rate=u)
+        else:
+            if length is not None:
+                raise ValueError("Cannot use sequence_length with named contig")
```

Comment:
`sequence_length` written here, rather than `length`. Maybe do a search for this and see if it's hiding anywhere else?